### DISCUSSION TITLE
Do you have any good ideas for block interaction effect?

### DISCUSSION THREAD

**Discussion by Sieluna** [DevBobcorn/CornCraft#22] [Ideas]
I initially planned to make a screen space effect, but I want the destruction effect to be even more refined. I'm considering using mesh splitting with multiple passes to achieve better outline and deformation effects.

@DevBobcorn What do you think?

**Comment by DevBobcorn** [comment]
I'd still like to use some vanilla-ish particles though. Textures for these particles are already ready for use, but their animation seems to be hardcoded in the vanilla game, so guess we'll need to work on that.

**Reply by DevBobcorn** [reply]
I'm also thinking implementing a custom particle system with ECS to improve performance, but I'm not sure exactly how.

**Reply by Sieluna** [reply]
To be honest, the destruction effect in MC isn’t that complex; we can just use a compute shader for it. The source generation compilations for entities are really slow.

**Reply by DevBobcorn** [reply]
What about block particles?

**Comment by DevBobcorn** [comment]
Oh did you mean the block cracking textures? I haven't really thought about it before.

**Reply by Sieluna** [reply]
I don't think this needs a texture at all

**Reply by DevBobcorn** [reply]
Yes the texture is not really needed, just mentioning this because vanilla did it this way

**Reply by DevBobcorn** [reply]
I wonder how what it's gonna look like, can you paste a preview of the visual effect or something like that?

**Reply by Sieluna** [reply]
![image](https://github.com/user-attachments/assets/133508fc-1287-4155-8748-5df153e9fe6d)
This is the ss effect

**Reply by DevBobcorn** [reply]
Looks cool, but why should this be a screen space effect? Shouldn't it be overlaid on block faces?

**Reply by Sieluna** [reply]
Good question. I just think this could be combined with the outline in screen space. ~~In fact, I just wanted to be lazy.~~

Actually, the current outline approach is pretty good. I just thought hard surface outline would be cool.

**Reply by DevBobcorn** [reply]
The outline always takes the shape of a full cube, but I just figured it would be better if the cracking textures can fit on the exact shape of the blocks like vanilla does it.

**Reply by DevBobcorn** [reply]
Or we can still make it a full cube, but instead of using cracking textures(which might look a bit wired if not fitting to the exact block mesh), we draw a progress bar on each side of the cube.

**Reply by DevBobcorn** [reply]
The progress bar idea actually came from oCd resource pack, but we don't necessarily need to make the progress bar look like this one, maybe something like the health bar we have.

![2024-11-13_23 53 20](https://github.com/user-attachments/assets/e6af9efb-d444-4c84-9589-c446982a3a85)

**Reply by Sieluna** [reply]
few days ago, I tried playing Minecraft and quickly realized that outside of Creative Mode, I felt like a complete beginner. I don’t think I've touched Survival Mode since 1.12. It’d be fun to try multiplayer sometime if there’s a chance!

**Comment by Sieluna** [comment]
Some inspiration in daydreaming, it looks better.

![image](https://github.com/user-attachments/assets/73cdcfe7-fd97-40aa-8279-d00508b33e0a)
![image](https://github.com/user-attachments/assets/facb6e5e-9dc7-442e-91a0-f66b014f01c7)

**Reply by DevBobcorn** [reply]
Looks nice, maybe we should try it. Do you have any idea about overlaying this effect onto block meshes?

**Comment by DevBobcorn** [comment]
I'm thinking about introducing Axis-Aligned Bounding Boxes for rendering cracking textures and selection boxes. They can also serve as more accurate and performant collision boxes (current mesh colliders can be slow to build).

**Comment by DevBobcorn** [comment]
Working on bounding box data, shouldn't take long to finish.

![image](https://github.com/user-attachments/assets/20372201-16c5-47ee-8f3f-0f5c2f1e22e7)
